crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+            tedge_config.device.cert_path.clone().into(),
+        )
+        .unwrap();
+        let (signal_sender, signal_receiver) = mpsc::channel(10);
+
+        let mut local_config = MqttOptions::new(
+            "tedge-mapper-bridge-c8y",
+            &tedge_config.mqtt.client.host,
+            tedge_config.mqtt.client.port.into(),
+        );
+        // TODO cope with secured mosquitto
+        local_config.set_manual_acks(true);
+        let mut cloud_config = MqttOptions::new(
+            tedge_config.device.id.try_read(tedge_config).unwrap(),
+            tedge_config
+                .c8y
+                .mqtt
+                .or_config_not_set()
+                .unwrap()
+                .host()
+                .to_string(),
+            8883,
+        );
+        cloud_config.set_manual_acks(true);
+
+        cloud_config.set_transport(Transport::tls_with_config(tls_config.into()));
+        let prefix = "c8y";
+        let (local_client, mut local_event_loop) = AsyncClient::new(local_config, 10);
+        let (cloud_client, mut cloud_event_loop) = AsyncClient::new(cloud_config, 10);
+
+        local_client
+            .subscribe("c8y/#", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/dt", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/dat", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/ds", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/e", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/dc/#", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("error", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+
+        let local_to_cloud = tokio::spawn(async move {
+            loop {
+                let notification = match cloud_event_loop.poll().await {
+                    Ok(notification) => notification,
+                    Err(err) => {
+                        error!("MQTT bridge connection error: {err}");
+                        continue;
+                    }
+                };
+                if let Event::Incoming(Incoming::Publish(publish)) = &notification {
+                    local_client
+                        .publish(
+                            format!("{prefix}/{}", publish.topic),
+                            publish.qos,
+                            publish.retain,
+                            publish.payload.clone(),
+                        )
+                        .await
+                        .unwrap();
+                    local_client.ack(publish).await.unwrap();
+                }
+            }
+        });
+        let cloud_to_local = tokio::spawn(async move {
```

Comment:
Instead of two independent loops, it would be better to have a single bridge logic that is used twice, with different mapping parameters.

I even wonder that these two loops might have to be merged with a `select!` over the two end-points. Indeed, one has to listen not only for messages from the source end-point but also for acknowledgements from the target end-point.

Reply:
That makes sense. I would be fearful of using `select!` here, it sounds like a recipe for recreating the bug in #2326, but it would make sense to limit the duplication of the logic code.